presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java

```diff
@@ -131,22 +145,62 @@ public void stop()
         return future;
     }
 
+    public synchronized ListenableFuture<?> waitForMinimumCoordinators()
+    {
+        if (currentCoordinatorCount >= minCoordinatorCount) {
+            return immediateFuture(null);
+        }
+
+        SettableFuture<?> future = SettableFuture.create();
+        coordinatorSizeFutures.add(future);
+
+        // if future does not finish in wait period, complete with an exception
+        ScheduledFuture<?> timeoutTask = executor.schedule(
+                () -> {
+                    synchronized (this) {
+                        future.setException(new PrestoException(
+                                GENERIC_INSUFFICIENT_RESOURCES,
+                                format("Insufficient active coordinator nodes. Waited %s for at least %s coordinators, but only %s coordinators are active", executionMaxWait, 2, currentCoordinatorCount)));
+                    }
+                },
+                maxCoordinatorWait.toMillis(),
+                MILLISECONDS);
+
+        // remove future if finished (e.g., canceled, timed out)
+        future.addListener(() -> {
+            timeoutTask.cancel(true);
+            removeFuture(future);
+        }, executor);
+
+        return future;
+    }
+
     private synchronized void removeFuture(SettableFuture<?> future)
     {
-        futures.remove(future);
+        workerSizeFutures.remove(future);
     }
 
     private synchronized void updateAllNodes(AllNodes allNodes)
     {
-        if (includeCoordinator) {
-            currentCount = allNodes.getActiveNodes().size();
-        }
-        else {
-            currentCount = Sets.difference(allNodes.getActiveNodes(), allNodes.getActiveCoordinators()).size();
-        }
+        currentCount = 0;
+        currentCoordinatorCount = 0;
+        allNodes.getActiveNodes().forEach(node -> {
```

Comment:
Any specific reason to not use the `getActiveNodes` and `getActiveCoordinator` methods here instead and again calculating it?

Reply:
That's a good point, thanks.